feat: add new format-entry function, note template

bdarcus opened the pull request:
Per #269, this adds a template for formatting new note titles, and a function to handle it.

It has occurred to me maybe this function should be called `format-entry`, and the original one renamed `format-entry-columns`?

Note: there's a [little change](https://github.com/bdarcus/bibtex-actions/pull/274/commits/6ddbe996b6d6d3c6b75379ac0ad7e80caa5b0c92#diff-43b58707c1adce5c6f3e1ed8d97ea5665182d94767472f7c196d630ff3ca0c3cL452) I made here that exposed a much bigger infinite recursion bug that I can't figure out how to fix otherwise.

```elisp
ELISP> (bibtex-actions--format-candidates bibtex-actions-bibliography)
*** Eval error ***  Lisp nesting exceeds ‘max-lisp-eval-depth’
```

Backtrace:

```elisp
Debugger entered--Lisp error: (error "Lisp nesting exceeds ‘max-lisp-eval-depth’")
  #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_17>("${author editor:30}")
  replace-regexp-in-string("\\$\\({\\([^}]+\\)}\\|[0-9]+\\)" #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_17> "${author editor:30}     ${date year issued:4}     ..." t t)
  s-format("${author editor:30}     ${date year issued:4}     ..." #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_20>)
  #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21>("${author editor:30}     ${date year issued:4}     ...")
  mapcar(#<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21> ("${author editor:30}     ${date year issued:4}     ..." "          ${=key= id:15}    ${=type=:12}    ${tags..."))
  #f(compiled-function #'sequence #<bytecode 0x18574107428dcbb4>)(#<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21> ("${author editor:30}     ${date year issued:4}     ..." "          ${=key= id:15}    ${=type=:12}    ${tags..."))
  apply(#f(compiled-function #'sequence #<bytecode 0x18574107428dcbb4>) #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21> ("${author editor:30}     ${date year issued:4}     ..." "          ${=key= id:15}    ${=type=:12}    ${tags...") nil)
  seq-map(#<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21> ("${author editor:30}     ${date year issued:4}     ..." "          ${=key= id:15}    ${=type=:12}    ${tags..."))
  seq-mapcat(#<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_21> ("${author editor:30}     ${date year issued:4}     ..." "          ${=key= id:15}    ${=type=:12}    ${tags..."))
  bibtex-actions--fields-in-formats()
  bibtex-actions--fields-to-parse()
  bibtex-actions--format-candidates(("/home/bruce/org/bib/newer.bib"))
```

bdarcus commented on bibtex-actions-file.el:
```diff
@@ -165,9 +167,13 @@ use 'orb-edit-note' for this value."
+             (bibtex-actions-get-template 'note)))
            (content
-            (concat ":PROPERTIES:\n:ID:  " uuid "\n:END:\n#+title: Notes on " title "\n")))
+            (concat ":PROPERTIES:\n:ID:  " uuid "\n:END:\n" note-title "\n")))
```
There's an issue here:

Right now, I have half of this which is handled via template, and the other via hard-coded function logic.

It might be better to allow the template a `uuid-new` or some such field, so the template could include of it:

```elisp
 ":PROPERTIES:\n:ID: ${uuid-new}\n:END:\n#+title: Notes on ${author editor}, ${title}"
```

bdarcus replied:
Maybe better to use something like this, to remove the `org-id` dependency?

http://ergoemacs.org/emacs/elisp_generate_uuid.html

aikrahguzar commented:
> Note: there's a [little change](https://github.com/bdarcus/bibtex-actions/pull/274/commits/6ddbe996b6d6d3c6b75379ac0ad7e80caa5b0c92#diff-43b58707c1adce5c6f3e1ed8d97ea5665182d94767472f7c196d630ff3ca0c3cL452) I made here that exposed a much bigger infinite recursion bug that I can't figure out how to fix otherwise.

Is it adding the `if` statement causing the infinite loop? That is very weird, because there is no recursion there at all. I checked by evaluating 

```elisp
(defun my-ba-get-entry (key)
  "Return the cached entry for KEY."
  (if (or (eq 'uninitialized bibtex-actions--candidates-cache)
          (eq 'uninitialized bibtex-actions--local-candidates-cache))
      (message "Something is wrong; your library is not initialized.")
    (cddr (seq-find
           (lambda (entry)
             (string-equal key (cadr entry)))
           (bibtex-actions--get-candidates)))))
```
and the calling it with a key and it seems to return the entry as expected. 

I am setting up at a new place so won't do much anytime soon but I will try to track down infinite recursion bug if I can reproduce it.

bdarcus commented:
> Is it adding the `if` statement causing the infinite loop?

No; the reverse. Adding that statement originally "fixed" it.

> I am setting up at a new place so won't do much anytime soon but I will try to track down infinite recursion bug if I can reproduce it.

OK.

bdarcus commented:
> Adding that statement, along with changing the load order of some functions, originally "fixed" it.

I added this check back (which again "fixed" it), but also added a "FIX" comment, because I don't think this check should be necessary if everything else is correctly coded.

aikrahguzar commented:
@bdarcus, does changing `ba--get-candidates` to
```elisp
(defun bibtex-actions--get-candidates (&optional force-rebuild-cache)
  "Get the cached candidates.
If the cache is unintialized, this will load the cache.
If FORCE-REBUILD-CACHE is t, force reload the cache."
  (when force-rebuild-cache
    (bibtex-actions-refresh force-rebuild-cache))
  (when (eq 'uninitialized bibtex-actions--candidates-cache)
    (bibtex-actions-refresh nil 'global))
  (when (eq 'uninitialized bibtex-actions--local-candidates-cache)
    (bibtex-actions-refresh nil 'local))
  (seq-concatenate 'list
                   bibtex-actions--local-candidates-cache
                   bibtex-actions--candidates-cache))
```
helps with the recursion blowing up? There is certainly a bug as currently the `'local` and `'global` checks are inside an `if` statement that won't be executed unless `force-rebuild-cache` is non-`nil`. Which it won't be in the normal first usage and that behavior would be fixed by checking for uninitialized variables for in `ba-get-entry`.

I don't see why it would cause infinite recursion instead of a type error though, unless it is something that happens during `completing-read` part. That seems unlikely to me but I don't understand that part well.

I would test and do a pr but I need to setup my laptop since the last one went with my last job.

bdarcus commented:
> does changing `ba--get-candidates` to ... helps with the recursion blowing up?

Alas, no.

aikrahguzar commented:
@bdarcus  I figured out the problem, but I don't have a good solution yet. It is the use of `bibtex-actions-file--files-for-key` in the `bibtex-actions-file--format-candidates`. `bibtex-actions-file--files-for-key` calls `bibtex-actions-file--possible-names` which calls `ba-get-entry` which calls `ba-get-candidates` and `ba-get-candidates` will call `bibtex-actions-file--format-candidates` because the cache isn't built yet causing bad recursion. The check for initialization then allow us to proceed.  This is a pretty subtle result of recent changes to parse the file field. I think the solution should be to change `ba-file` functions to take the extra `entry` argument and pass the entry directly for `ba--format-candidates` and use the `ba-get-entry` in interactive functions that need the files to supply this new entry argument.

This has me thinking that `ba-get-entry` exists only to overcome a shortcoming in the `ba-select-keys`. We store the entry but then throw it away when we return only the keys in `ba-select-keys` and then we need to go through the whole candidate list to find the entry later when we need it. It is inefficient but that doesn't really matter since we only need to lookup a few selected keys but it can also cause this kind of issue. We can change `ba-select-keys` to return an alist of `(key . entry)` but that will probably need some adjustment for embark integration too.

bdarcus commented:
Maybe get-entry needs to grab from the candidate cddr then?

If that's possible ...

aikrahguzar commented:
> Maybe get-entry (or just however to get the necessary data) needs to grab from the candidate cddr then? If that's possible; not sure it is.

Something like this would be my preferred solution. Where `select-keys` returns alist but a lot of (I think all the interactive) functions would needed to be update to deal with this change. Though the change would just be to change the `key` to `key-entry` and then get the key by  (car key-entry). The other problem is embark's at point and we would need to change that case to fetch the entry like we do in `get-entry` now. That is the only place where I think the current solution is the optimal one.

bdarcus commented:
Looks like the only field files-for-key is currently accessing (indirectly) is `files`.

aikrahguzar commented:
> Looks like the only field files-for-key is currently accessing (indirectly) is `files`.

Yes, but I think generally the set of functions that depend on the caches being populated should be fairly limited, I guess ideally only to interactive functions.

aikrahguzar commented:
> Looks like the only field files-for-key is currently accessing (indirectly) is `files`.

Yes, but I think generally the set of functions that depend on the caches being populated should be fairly limited, I guess ideally only to interactive functions.

bdarcus commented:
One issue, the details of which I'm forgetting such that I'm unsure if it applies here, is that completing-read throws out non-candidate data at a certain point.

bdarcus commented:
Occurs to me also that it might be wise at some to add some tests, that we can also hook up to the CI.

aikrahguzar commented:
> One issue, the details of which I'm forgetting such that I'm unsure if it applies here, is that completing-read throws out non-candidate data at a certain point.

I think we already take care of it to find the key corresponding to the formatted string. I was thinking of changing this
```elisp
(cl-loop for choice in chosen
             ;; Collect citation keys of selected candidate(s).
             collect (or (cadr (assoc choice candidates))
             ;; Key is literal coming from embark, just pass it on
                          choice))
```
to something like this
```elisp
(seq-map (lambda (choice) (seq-find
                            (lambda (cand) (cdr (or (equal choice (car cand))
                                             ;; For embark, find the cand with correct key
                                                    (equal choice (cadr cand)))))
                          candidates))
             chosen)
```
Basically change `cadr` to `cdr` so that the function calling `select-keys` can use both the key and entry.

bdarcus commented:
> Basically change `cadr` to `cdr` so that the function calling `select-keys` can use both the key and entry.

I pushed #274 before seeing this, but will take a look sometime today or tomorrow.

Are you thinking of this as a simpler alternative to #274?

aikrahguzar commented:
> I pushed #274 before seeing this, but will take a look sometime today or tomorrow.
> 
> Are you thinking of this as a simpler alternative to #274?

I briefly looked at #274 and it looks good. This is almost unrelated and the purpose is not to go through candidate list too many times. I think it cleaner if we have a separation of the part of the library that does caching and the rest of the library interacts with cached candidates only through `select-keys`.

bdarcus commented:
> I think it cleaner if we have a separation of the part of the library that does caching and the rest of the library interacts with cached candidates only through `select-keys`.

Yes, ideally they really should be cleanly-separated.

bdarcus commented:
> I pushed #274 before seeing this ...

Sigh ... I of course meant #281!
